---
title: "Assigning ASVs with DADA2"
author: "Mark Watson
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/01_DADA2")
```

# Symbolically link out data files

We need to have access to our fastq files for our project

# Goals of this file:

1. Load in raw seq. data
2. Visualize/inspect seq. data to assess read quality
3. Filter by quality and trim  primers from seq. data
4. Wrute out new fastq files w/ high quality and trimmed sequences
5. Reinspect quality of new reads (intuition check)
6. Infer the error rate of forward (R1) and reverse (R2) reads
**DADA2 building the error model**
7. Apply error model and infer ASVs on F and R reads individually
8. Merge forward and reverse ASVs into "contiguous ASVs"
9. Generate an ASV count table ('otu_table` for phyloseq.)

Output that we need:

1. ASV count table: `otu_table`
2. 
3. 
