---
title: "Assigning ASVs with DADA2"
output: html_document
date: "2025-03-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = 'center',
                      #send figures generated to folder below
                      fig.path = "../figures/02_AssignASVs/")
```


# Goals
1. Infer Errors in our sequences, separately on F and R reads.
2. Assign ASVs on both F and R reads. (apply error model)
3. Merge F and R ASVs into "contiguous ASVs".
4. Generate first draft of ASV count table.
5. Quality Trimming of ASV lengths. 
6. Remove chimeras. 
7. Assign Taxonony w/ Silva Database.
8. Write out relevant files : `asv_table`, `asvs_fasta`, `tax_table`, and `sample_data`

## Input
1. Filtered Fastq files generatedf from `01_QualityTrimming.Rmd`.
2. Sample Name vector.


## Output
1. `asv_table`
2. `asvs_fasta`
3. `tax_table`
4. `sample_data`

# Set up the Environment
## Set Seed
```{r set-seed}
set.seed(238428)

```
## Load Packages
```{r load-packages}
pacman::p_load(tidyverse, devtools, dada2,
               patchwork, DT, install = FALSE)
```
## Load Filtered Fastq Files
```{r load-filtered-fastqs}
#place filtered seqs into variable
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"
#intuition check
filtered_fastqs_path

#Create F and R vector
filtered_forward_reads <- 
  list.files(filtered_fastqs_path, pattern = "R1_filtered.fastq.gz", full.names = TRUE)
filtered_reverse_reads <- 
  list.files(filtered_fastqs_path, pattern = "R2_filtered.fastq.gz", full.names = TRUE)
#Check
filtered_forward_reads[1:5]
filtered_reverse_reads[1:5]

```
## Sample Names
```{r sample-names}
# Create vector of sample names from the filenames 
sample_names <- sapply(strsplit(basename(filtered_forward_reads), "_"), `[`,1) 

# Intuition Check 
head(sample_names)
```
# Error Modeling
```{r learn-errors}
#Forward Reads
error_forward_reads <-
  learnErrors(filtered_forward_reads, multithread = 6)

#Plot
forward_error_plot <-
  plotErrors(error_forward_reads, nominalQ = TRUE)

#Reverse Reads
error_reverse_reads <-
  learnErrors(filtered_reverse_reads, multithread = 6)
  
#Plot
reverse_error_plot <-
  plotErrors(error_reverse_reads, nominalQ = TRUE)+
  labs(title = "Reverse Reads:Error Model")

#look @ plots
forward_error_plot + reverse_error_plot
```
# Infer sequencing reads
```{r infer-asvs}
#dada Forward ASVs
dada_forward <- dada(filtered_forward_reads, 
                     err = error_forward_reads, 
                     multithread = 6)
# dada reverse ASVs
dada_reverse <- dada(filtered_reverse_reads, 
                     err = error_reverse_reads,
                     multithread = 6)

# take a look at data
typeof(dada_forward)
typeof(dada_reverse)
 # dada_forward$`<sample id here>`

```

#Merge F and R ASVs
```{r merge-ASVs}
merged_ASVs <-
  mergePairs(
    dada_forward, filtered_forward_reads,
    dada_reverse, filtered_reverse_reads, 
    verbose = TRUE)
```
# Evaluate the data output
typeof(merged_ASVs)
length(merged_ASVs)
merged_ASVs

# Create Raw ASV Count Table
```{r raw-ASV-count-table}
#Raw ASV
raw_ASV_table <-
  makeSequenceTable(merged_ASVs)

#check
dim(raw_ASV_table)
typeof(raw_ASV_table)
class(raw_ASV_table)

#write out raw_asv_table
write.table(raw_ASV_table, file = "data/01_DADA2/raw_ASV_counts.tsv",
            sep = "\t", quote = FALSE, col.names = NA)
```
# Assess the ASV Length
Some more QC
```{r assess-ASV-length}

#Creating a table to inspect the distribution of ASV length
getSequences(raw_ASV_table) %>% #create list of all ASVs
  nchar() %>% #count characters of each item in list
  table() # show table of ^^^

# Plot
data.frame(ASV_Length = nchar(getSequences(raw_ASV_table))) %>%
  ggplot(aes(x = ASV_Length)) +
  geom_histogram() +
  scale_x_continuous(limits = c(0,500))+
  labs(title = "Raw ASV Lengths",
       y = "Number of ASVs", x = "ASV Sequence Length(bps)")

```

# Trim the ASVs
```{r trim-ASVs}
# only take ASVs == 245 bp
raw_ASV_table_trimmed <-
  raw_ASV_table[, nchar(getSequences(raw_ASV_table)) == 245]

#intuition check
table(nchar(getSequences(raw_ASV_table_trimmed)))

```

# Remove Chimeras
```{r rm-chimeras}
# Remove the chimeras in the raw ASV table
noChimeras_ASV_table <- 
  removeBimeraDenovo(raw_ASV_table_trimmed, 
                    method="consensus", 
                    multithread=TRUE, verbose=TRUE)

# What proportion is left of the sequences? 
percRetained_chimerasTrimmed <- sum(noChimeras_ASV_table)/sum(raw_ASV_table_trimmed)
percRetained_chimerasRaw <-sum(noChimeras_ASV_table)/sum(raw_ASV_table)

maxLength_ASV <- max(nchar(getSequences(raw_ASV_table))) # Longest ASV?


# Plot it 
data.frame(Seq_Length_NoChim = nchar(getSequences(noChimeras_ASV_table))) %>%
  ggplot(aes(x = Seq_Length_NoChim )) + 
  geom_histogram()+ 
  # include the x-axis scales
  scale_x_continuous(limits = c(0, maxLength_ASV + 5)) + 
  labs(title = "Trimmed + Chimera Removal distribution of ASV length",
       y = "Number of ASVs", x = "ASV Sequence Length (bps)")


```
# Track the number of reads DADA2 workflow

# Assign Taxonomy


# Session Information
```{r session-info}
#ensure reproducibility with package version info
devtools::session_info()
```
